components/resource_control/src/service.rs

```diff
@@ -139,20 +152,215 @@ impl ResourceManagerService {
+                            let _ = GLOBAL_TIMER_HANDLE
+                                .delay(std::time::Instant::now() + RETRY_INTERVAL)
+                                .compat()
+                                .await;
+                            continue;
+                        }
+                    }
+                }
+                Err(err) => {
+                    error!("failed to load controller config"; "err" => ?err);
+                    let _ = GLOBAL_TIMER_HANDLE
+                        .delay(std::time::Instant::now() + RETRY_INTERVAL)
+                        .compat()
+                        .await;
+                    continue;
+                }
+            }
+        }
+    }
+
+    // upload ru metrics periodically.
+    pub async fn upload_ru_metrics(&self) {
+        let mut last_group_statistics_map: HashMap<String, UploadStatistic> = HashMap::new();
+        // load controller config firstly.
+        let config = self.load_controller_config().await;
+        info!("load controller config"; "config" => ?config);
+
+        loop {
+            let background_groups: Vec<_> = self
+                .manager
+                .resource_groups
+                .iter()
+                .filter_map(|kv| {
+                    let g = kv.value();
+                    g.limiter.clone().map(|limiter| {
+                        let io_statistics = limiter.get_limit_statistics(ResourceType::Io);
+                        let cpu_statistics = limiter.get_limit_statistics(ResourceType::Cpu);
+
+                        (
+                            g.group.name.clone(),
+                            UploadStatistic {
+                                // io statistics and cpu statistics should have the same version.
+                                version: io_statistics.version,
+                                read_bytes_consumed: io_statistics.read_consumed,
+                                write_bytes_consumed: io_statistics.write_consumed,
+                                cpu_consumed: cpu_statistics.total_consumed,
+                            },
+                        )
+                    })
+                })
+                .collect();
+
+            if background_groups.is_empty() {
+                let _ = GLOBAL_TIMER_HANDLE
+                    .delay(std::time::Instant::now() + BACKGROUND_RU_UPLOAD_DURATION)
+                    .compat()
+                    .await;
+                continue;
+            }
+
+            let mut req = TokenBucketsRequest::default();
+            let all_reqs = req.mut_requests();
+            for (name, statistic) in background_groups.into_iter() {
+                // check if the entry exists in the map or create a new one with u64::MAX
+                // version to avoid losing the first upload.
+                let last_stats = last_group_statistics_map
+                    .entry(name.clone())
+                    .or_insert_with(|| UploadStatistic::new(u64::MAX));
```

Comment:
why explicitly set u64::MAX here, I think the default value 0 should be enough.

Reply:
This is a flag to check if it is the first time it is stored in `last_group_statistics_map`... Let me see if there is a better way to distinguish between `equal` and `new insert`.

Reply:
fixed it by checking the existence and version, PTAL again! thx! @glorv